src/main/java/appeng/util/item/UnsortedItemList.java

```diff
@@ -0,0 +1,165 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Applied Energistics 2.  If not, see <http://www.gnu.org/licenses/lgpl>.
+ */
+
+package appeng.util.item;
+
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterators;
+
+import appeng.api.config.FuzzyMode;
+import appeng.api.storage.data.IAEItemStack;
+import appeng.api.storage.data.IItemList;
+
+
+/**
+ * An unsorted {@link IItemList} providing constant access time instead of logarithmic time.
+ *
+ * As tradeoff it will no longer support fuzzy operations.
+ * Also no advanced features like storing craftable or requestable items is supported.
+ *
+ */
+public final class UnsortedItemList implements IItemList<IAEItemStack>
+{
+
+	private static final Predicate<IAEItemStack> MEANINGFUL_PREDICATE = new Predicate<IAEItemStack>(){
+
+		@Override
+		public boolean apply( IAEItemStack input )
+		{
+			return input.isMeaningful();
+		}
+	};
+
+	private final Map<IAEItemStack, IAEItemStack> records = new HashMap<IAEItemStack, IAEItemStack>();
+
+	@Override
+	public void add( IAEItemStack option )
+	{
+		if( option == null )
+		{
+			return;
+		}
+
+		final IAEItemStack st = this.records.get( option );
+
+		if( st != null )
+		{
+			st.add( option );
+			return;
+		}
+
+		final IAEItemStack opt = option.copy();
+
+		this.records.put( opt, opt );
+	}
+
+	@Override
+	public IAEItemStack findPrecise( IAEItemStack itemStack )
+	{
+		if( itemStack == null )
+		{
+			return null;
+		}
+
+		return this.records.get( itemStack );
+	}
+
+	@Override
+	public boolean isEmpty()
+	{
+		return this.records.isEmpty();
+	}
+
+	@Override
+	public int size()
+	{
+		return this.records.size();
+	}
+
+	@Override
+	public Iterator<IAEItemStack> iterator()
+	{
+		return Iterators.filter( this.records.values().iterator(), MEANINGFUL_PREDICATE );
+	}
+
+	@Override
+	public IAEItemStack getFirstItem()
+	{
+		for( IAEItemStack stackType : this )
+		{
+			return stackType;
+		}
+
+		return null;
+	}
+
+	@Override
+	public void resetStatus()
+	{
+		for( IAEItemStack i : this )
+		{
+			i.reset();
+		}
+	}
+
+	/**
+	 * Unsupported due to being a unsorted collection and thus only solvable in linear or worse time.
+	 */
+	@Override
+	@Deprecated
+	public Collection<IAEItemStack> findFuzzy( IAEItemStack filter, FuzzyMode fuzzy )
+	{
+		throw new UnsupportedOperationException( "Unsupported on an unsorted collection" );
+	}
+
+	/**
+	 * Unsupported to avoid being used as anything but a plain collection to store real itemstacks.
+	 */
+	@Override
+	@Deprecated
+	public void addStorage( IAEItemStack option )
+	{
+		throw new UnsupportedOperationException( "Purely designed for item storage" );
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "Purely designed for item storage" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1192)